src/python/zquantum/core/wavefunction.py

```diff
@@ -0,0 +1,142 @@
+from math import log2
+from typing import Any, Dict, List, Union
+from warnings import warn
+
+import numpy as np
+from sympy import Abs, Matrix, Symbol
+
+
+def _cast_sympy_matrix_to_numpy(sympy_matrix, complex=False):
+    new_type = np.complex128 if complex else np.float64
+
+    try:
+        return np.array(sympy_matrix, dtype=new_type).flatten()
+    except TypeError:
+        return np.array(sympy_matrix, dtype=object).flatten()
+
+
+class Wavefunction:
+    """
+    A simple wavefunction data structure that can
+    be used to calculate amplitudes of quantum states.
+
+    Args:
+        amplitude_vector: the initial amplitudes of the system,
+            can either be a NumPy ndarray or a SymPy Matrix
+    """
+
+    def __init__(
+        self, amplitude_vector: Union[np.ndarray, Matrix, List[complex]]
+    ) -> None:
+        if bin(len(amplitude_vector)).count("1") != 1:
+            raise ValueError(
+                "Provided wavefunction does not have a size of a power of 2."
+            )
+
+        if isinstance(amplitude_vector, np.ndarray) or isinstance(
+            amplitude_vector, list
+        ):
+            amplitude_vector = Matrix(amplitude_vector)
+
+        self._check_sanity(amplitude_vector)
+
+        self._wavefunction = amplitude_vector
```

Comment:
Why not actually call it `_amplitude_vector` instead of`_wavefunction` ?

Reply:
@mstechly I am just trying to avoid using the exact naming as PyQuil. If you think the naming is weird, I can change it

Reply:
Please change :)